File: app/src/main/java/org/deiverbum/app/viewmodel/BibliaViewModel.java
package org.deiverbum.app.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

import org.deiverbum.app.data.wrappers.CustomException;
import org.deiverbum.app.data.wrappers.DataWrapper;
import org.deiverbum.app.model.BibleBooks;
import org.deiverbum.app.repository.BibliaRepository;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;

/**
 * ViewModel para el módulo Biblia
 * @author A. Cedano
 * @since 2022.01.01
 */
@HiltViewModel
public class BibliaViewModel extends ViewModel {
    private final BibliaRepository mRepository;

@Inject
    public BibliaViewModel(BibliaRepository repository) {
        mRepository = repository;
    }

    public LiveData<DataWrapper<BibleBooks, CustomException>> getLibro(int param) {
        return mRepository.getLibro(param);
    }
}

File: app/src/main/java/org/deiverbum/app/presentation/biblia/BibliaViewModel.kt
package org.deiverbum.app.presentation.biblia

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import org.deiverbum.app.data.networking.CoroutineDispatcherProvider
import org.deiverbum.app.domain.model.BibleBookRequest
import org.deiverbum.app.domain.usecase.GetBibleBookUseCase
import javax.inject.Inject

/**
 * ViewModel para la sincronización.
 *
 * @author A. Cedano
 * @version 2.0
 * @since 2022.2
 */
@HiltViewModel
class BibliaViewModel @Inject constructor(
    private val getBibleBookUseCase: GetBibleBookUseCase,
    private val coroutineDispatcherProvider: CoroutineDispatcherProvider
) : ViewModel() {

    private val _uiState = MutableStateFlow<BibliaUiState>(BibliaUiState.Empty)
    val uiState: StateFlow<BibliaUiState> = _uiState

    /**
     * Este método se ocupa de buscar el libro de la Biblia.
     *
     * @param bookRequest Es un objeto [BibleBookRequest] con toda la información de la sincronización que se está solicitando.
     */
    fun loadData(bookRequest: BibleBookRequest) {
        _uiState.value = BibliaUiState.Loading
        viewModelScope.launch(coroutineDispatcherProvider.IO()) {
            try {
                val result = getBibleBookUseCase.execute(bookRequest)
                _uiState.value = BibliaUiState.Loaded(BibliaItemUiState(result))
            } catch (error: Exception) {
                _uiState.value = BibliaUiState.Error(error.message.toString())
            }
        }
    }



    sealed class BibliaUiState {
        object Empty : BibliaUiState()
        object Loading : BibliaUiState()
        class Loaded(val itemState: BibliaItemUiState) : BibliaUiState()
        class Error(val message: String) : BibliaUiState()
    }
}